torch/csrc/autograd/engine.cpp

```diff
@@ -386,16 +386,18 @@ void Engine::thread_on_exception(
 void GraphTask::set_exception(
     std::exception& e,
     const std::shared_ptr<Node>& fn) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::unique_lock<std::mutex> lock(mutex_);
   if (!has_error_.load()) {
     if (AnomalyMode::is_enabled() && fn) {
       fn->metadata()->print_stack();
```

Comment:
Thanks for the detailed explanation, although can't we have the same problem with this patch? The GraphTask can be deallocated here, before we unlock right?

Shouldn't the fix be something like what I've done in https://github.com/pytorch/pytorch/pull/33214, where we hold the lock before calling `set_exception` in the callback for `addOutstandingRpc`?

Reply:
Not sure I see how GraphTask can be deallocated in print_stack()?  (I admit that I'm newish to this module)
The cases that I observed in particular involved the future callback triggering the dist context to clear its current task - which this avoids by dropping the lock prior to the potentially triggered delete.

btw, aside, re: #33214 that you referenced - can you possibly avoid changing the context's std::mutex to std::recursive_mutex?